### PR TITLE
PL-397 Add microseconds to the database name to avoid naming clashes

### PR DESCRIPTION
**Pull request by AlexHayton**
https://homeservenow.atlassian.net/browse/PL-397

## Brief description of what's changing
- [x] Add microseconds to the database name in a backwards-compatible way
- [x] Improve test coverage

## Brief description of why it's changed
To avoid duplicate DB name errors when running lots of tests at around the same time.

### CODE REVIEW

**Review by mthomasuk — APPROVED** [review]
🎉

File: src/sequelize.init.ts [hunk 1]
```diff
@@ -27,26 +27,19 @@ const initialize = (props?: InitializeProps): Sequelize => {
   });
 };
 
-export class SequelizeSingleton {
-  private static instance: Sequelize | null;
+let instance: Sequelize | null;
```
Comment (alecdotico) [inline src/sequelize.init.ts:30]: This was explicitly implemented as OO to state the singleton pattern. This functional approach might produce [the same](https://stackoverflow.com/a/47298363/2586601) output as the OO approach, but I am not entirely sure, would have to investigate further into Javascript import evaluation. 
Anyway, for future readers it won't be as clear that this is a singleton pattern

File: src/testSuite/testHelpers.ts [hunk 2]
```diff
@@ -31,14 +32,15 @@ type NewTestDatabase = {
...
+export const microseconds = (): number => process.hrtime()[1];
 
 const dbSuffixToDate = (suffix: string): Date | null => {
   const date = parse(suffix, DATE_FORMAT, new Date());
   return isValid(date) ? date : null;
 };
 
 const newTestDatabase = (existingDbName: string): NewTestDatabase => {
-  const tempDbName = `${existingDbName}_test${dateToDbSuffix(new Date())}`;
+  const tempDbName = `${existingDbName}_${microseconds()}_test${dateToDbSuffix(new Date())}`;
```
Comment (alecdotico) [inline src/testSuite/testHelpers.ts:43]: microseconds should be part of the timestamp suffix, inside the `dateToDbSuffix` function. For instance, `jobs_test2021021515652001`. Thus, all temporary databases can be sorted alphabetically and deleted by timestamp.

Comment (alecdotico) [reply src/testSuite/testHelpers.ts:43]: this change also requires to update the `deleteOldTestDatabases` function, `'%_test[0-9]{15,18}'`

**Review by alecdotico — CHANGES_REQUESTED** [review]

File: src/sequelize.init.ts (hunk 1 above)
Comment (AlexHayton) [reply src/sequelize.init.ts:30]: I actually found your one harder to read! I guess I'm just used to doing singletons this way... @mthomasuk casting vote?

File: src/testSuite/testHelpers.ts (hunk 2 above)
Comment (AlexHayton) [reply src/testSuite/testHelpers.ts:43]: ok I'll refactor this

Comment (AlexHayton) [reply src/testSuite/testHelpers.ts]: Done. @alecdotico please can you take another look?

File: src/testSuite/testHelpers.ts [hunk 3]
```diff
@@ -29,18 +29,21 @@ type NewTestDatabase = {
...
 
 const dbSuffixToDate = (suffix: string): Date | null => {
-  const date = parse(suffix, DATE_FORMAT, new Date());
+  const date = parse(suffix.substring(0, 12), DATE_FORMAT, new Date());
   return isValid(date) ? date : null;
 };
 
 const newTestDatabase = (existingDbName: string): NewTestDatabase => {
-  const tempDbName = `${existingDbName}_${microseconds()}_test${dateToDbSuffix(new Date())}`;
+  const tempDbName = `${existingDbName}_test${dateToDbSuffix(new Date())}${microseconds()}`;
```
Comment (alecdotico) [inline src/testSuite/testHelpers.ts:46]: Looking better, but I still think that the `microseconds` function should be an internal of the `dateToDbSuffix` function. See it as a random seed added to the end of the suffix.

**Review by alecdotico — APPROVED** [review]

File: src/testSuite/testHelpers.ts (hunk 3 above)
Comment (AlexHayton) [reply src/testSuite/testHelpers.ts:46]: Alright! :)

File: src/sequelize.init.ts (hunk 1 above)
Comment (mthomasuk) [reply src/sequelize.init.ts:30]: it's harder to read but yeah, who fucking knows how JS is going to handle this - seems to be working in the current format, maybe leave it? unless someone wants to go full metal jacket with this

**Pull request merged**